src/ism/manifest.rs

```diff
@@ -0,0 +1,151 @@
+use std::io::{Read, Write};
+
+use yaserde::{YaDeserialize, YaSerialize};
+
+use crate::utils;
+
+#[derive(Debug, YaDeserialize, YaSerialize)]
+#[yaserde(root = "smil", namespace = "http://www.w3.org/2001/SMIL20/Language")]
+pub struct Smil {
+  head: Head,
+  body: Body,
+}
+
+impl Smil {
+  pub fn get_audio_stream_sources(&self) -> Vec<String> {
+    utils::remove_duplicates(
+      self
+        .body
+        .switch
+        .audio
+        .iter()
+        .map(|audio| audio.source.clone())
+        .collect(),
+    )
+  }
+
+  pub fn get_video_stream_sources(&self) -> Vec<String> {
+    utils::remove_duplicates(
+      self
+        .body
+        .switch
+        .video
+        .iter()
+        .map(|video| video.source.clone())
+        .collect(),
+    )
+  }
+
+  pub fn get_text_stream_sources(&self) -> Vec<String> {
+    utils::remove_duplicates(
+      self
+        .body
+        .switch
+        .text
+        .iter()
+        .map(|text| text.source.clone())
+        .collect(),
+    )
+  }
+}
+
+#[derive(Debug, YaDeserialize, YaSerialize)]
+struct Head {
+  meta: Vec<Meta>,
+  #[yaserde(rename = "paramGroup")]
+  param_group: Vec<ParamGroup>,
+}
+
+impl Default for Head {
```

Comment:
you can add this default to `derive` directly

Reply:
See 2578ac7